### ISSUE TITLE
Question related to typings for additional otherProps using interface merging

### ISSUE THREAD

**Issue by Silox** [ant-design/ant-design#4230]
English speaker, translated the template to english.

#### The environment in which the problem occurs is:

- antd Version:  2.5.2
- TypeScript version: 2.1.4
- Operating system and version: Windows 10
- Browser and its version: Not related to browser

#### What did you do? Please provide as detailed a step as possible.

I'm trying to add additional props to an <Input />, such as `required` and `form`, in a TypeScript environment, for example:

```
import { Input } from "antd";

<antd.Input form={formClass} value={value} required={true} onChange={e => this.handleChange(e)} />
```

#### The results you expect are:

I expected to be able to extend the InputProps interface to include the additional props so the TypeScript could compile without errors.

#### The actual result is:

The following errors are thrown while compiling:

```
Property 'form' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<Input> & { children?: ReactNode; } & InputProps'.at line 50 col 29
Property 'required' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<Input> & { children?: ReactNode; } & InputProps'.at line 52 col 29
```

This seems logical to me as the `InputProps` (https://github.com/ant-design/ant-design/blob/master/components/input/Input.tsx#L35) does not provide the `required` and `form` props, yet it allows for additional props to be passed:  https://github.com/ant-design/ant-design/blob/master/components/input/Input.tsx#L224. 

#### A reproducible online demo

Coulnd't provide a reproducible online demo as codepen does not seem to typecheck the typescript.

#### Tl;dr 

Is there a way to extend the InputProps from the `antd` library so I can add additional props? Or is there a better way to achieve this (without interface merging perhaps)?

**Comment by bang88** [comment]
@Silox If you need extends some extra props.  a) Wrap an `Input` component yourself and extends the antd `InputProps` interface. b) `antd` Team add a  string index signature like this 

```ts
export interface InputProps {
    prefixCls?: string;
    className?: string;
    type?: string;
    id?: number | string;
    value?: any;
    defaultValue?: any;
    placeholder?: string;
    size?: 'large' | 'default' | 'small';
    ...
    [propName: string]: any;
}
```

**Comment by Silox** [comment]
@bang88 Thank you for your quick reaction! I however don't see how I can implement solution a). I've tried to wrap my own ExtendedInput component using:

```
import { Input } from "antd";
import { InputProps } from "antd/lib/input/input";

interface ExtendedInputProps extends InputProps {
    form?: any;
}

export class ExtendedInput extends React.Component<ExtendedInputProps, any>  {
    render() {
        return(<Input form={this.props.form} {...this.props} />) // form still isn't a valid prop
    }
}

...

<ExtendedInput
    form={formClass} // form is now valid
    value={value}
    required={true}
    onChange={e => this.handleChange(e)}
/>
```

It results in the same situation as I still have to pass `form` to the antd Input field; this only moves the problem.

 I also tried it by extending the `Input` class and implementing the exteneded `InputProps` interface:

```
import { Input } from "antd";
import { InputProps } from "antd/lib/input/input";

interface ExtendedInputProps extends InputProps {
    form?: any;
}

export class ExtendedInput extends Input implements ExtendedInputProps  {
}

...

import { ExtendedInput } from "typings/antd/antd-additional";

<ExtendedInput
    form={formClass}
    value={value}
    required={true}
    onChange={e => this.handleChange(e)}
/>

``` 

But this yields the error `Property 'form' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<ExtendedInput> & { children?: ReactNode; } & Input...`

Am I missing something? Should I use HOC to achieve the extension? 

I like option b) as well as a the `extraProps` is already supported nicely by the components themselves and it would be less of a hassle to apply option a) for all concepts needing extension.

**Comment by Silox** [comment]
Right, I found a solution using HOC. I defined my interfaces as:
```
interface FormedProps {
    form?: any;
}

interface RequiredProps {
    required?: any;
}

interface IDAbleProps {
    id?: any;
}
```

And wrote 3 HOC functions:
```
export function Formed<T>(FormedTemplate: React.ComponentClass<T>): React.ComponentClass<T & FormedProps> {
    return class extends React.Component<T & FormedProps, void> {
        render() {
            return (
                <FormedTemplate {...this.props} />
            )
        }
    }
}

export function Required<T>(FormedTemplate: React.ComponentClass<T>): React.ComponentClass<T & RequiredProps> {
    return class extends React.Component<T & RequiredProps, void> {
        render() {
            return (
                <FormedTemplate {...this.props} />
            )
        }
    }
}

export function IDAble<T>(FormedTemplate: React.ComponentClass<T>): React.ComponentClass<T & IDAbleProps> {
    return class extends React.Component<T & IDAbleProps, void> {
        render() {
            return (
                <FormedTemplate {...this.props} />
            )
        }
    }
}
```

Now I can create my own components using:
```
const RequiredFormedInput = Required(Formed(Input));
const FormedButton = Formed(Button);
const IDAbleForm = IDAble(Form);
```

So I can use `<RequiredFormedInput required=string form=string ... />`,  `<FormedButton form=string />` and `<IDAbleForm id=string ... />`.

I have however no idea if there is a better solution, so better proposals are very welcome. If this is the correct way, I found a satisfying answer to this issue.

**Comment by bang88** [comment]
@Silox I have a little confuse, What's your use case ? why are you need these Props,are you using them for validate purpose ? If so you can take a look at the form docs https://ant.design/components/form/ .

**Comment by bang88** [comment]
And if you want extend the Input component,its not difficult:

```ts

import { Input } from "antd";

// input has no default export it's Props,
// if you want give it a try,you can import it directly.
import { InputProps } from "antd/lib/input/input";

// Your Input Props
export interface MyInputProps extends InputProps {
    required?: boolean;
    form?: any;
}
// Extends antd's Input
class MyInput extends React.Component<MyInputProps, any>{
    render() {
        return <Input {...this.props} />
    }
}
// Usage
<MyInput size="large" form="your form data can be any type" required={true} />

```

If you have any questions, please reopen this issue.

**Comment by Silox** [comment]
Hello @bang88,

Thanks again for your answer. The extending code you provided is indeed a lot more simple than my HOC solution.

The reason I'm not using the form validation is because I am displaying my data inside a table, more specifically the Editable Table example on https://ant.design/components/table/#components-table-demo-edit-row. As wrapping `<tr>` tags inside a `<form>` isn't valid HTML, I don't think I can make use of the `<Form />` component `antd` provides. (Correct me if I'm wrong, if there is a way to combine `<Form />` with `<Table />` where each row is its own form, it would be much cleaner. 

So at the moment, it's a trade-off to use the Editable Table from the documentation where sorting and filtering are provided, or use a solutions with `<div />`s  and `<Col />`s and benefit from `<Form />`'s validations.

**Comment by bang88** [comment]
@Silox Get it. Are the extend Input can solve your problem ?

**Comment by bang88** [comment]
You can render `Form` in a `Table` row.Its ok,if not please let me know. and if you still have issue please provide an online version.

**Comment by Silox** [comment]
@bang88: The extended Input does indeed solve my TypeScript problem, thank you!

I'm not sure what you mean by "You can render a `Form` in a `Table` row". Does the antd `Table` provide a function to render row-based instead of column-based?

What I actually want to achieve is the layout from the editable rows example (https://ant.design/components/table/#components-table-demo-edit-row) where I have 1 save-callback handler when a row is saved. This allows me to easily communicate the changed data of 1 row back to an API. Now the `handleChange()` code is called for every input field in the row. 

The ideal solution would be to have the following code:
```
<table>
  <thead>...</thead>
  <tbody>
    <form onSubmit=handleFormSubmit()
      <tr>
        <td><input>Name</input></td>
        ... 
        <td><button type="submit" /></td>
      </tr>
    </form>
  </tbody>
</table>
```

This allows for form validation using the antd `getFieldDecorators()` and have 1 save action for 1 row in the table so I can send the changed information in 1 time to the remote API. The problem with this however is that this isn't valid HTML, nor do I see a way to achieve this with the antd `Form` and `Table` components.

**Comment by sleva13** [comment]
@Silox did you find a way to add Form validation to Table row?

**Comment by Silox** [comment]
@sleva13 No, I decided to not check anything clientside and let my API handle all user validation. It already did before and it's less immediate and less user-friendly than clientside validation, but deadlines were coming so it was the quickest way.

**Comment by sleva13** [comment]
@Silox thanks for your quick reply. I will do the same.
